### PR TITLE
[primTorch] Add ref for `huber_loss` and error inputs

### PR DESCRIPTION
**Pull request by nkaretnikov**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #85041

### CODE REVIEW

File: torch/_refs/nn/functional/__init__.py [hunk 1]
```diff
@@ -419,6 +420,32 @@ def hinge_embedding_loss(
...
+    """
+    Reference implementation of torch.nn.functional.huber_loss
+    """
+    _check_reduction_value(reduction)
+    check(
+        delta > 0,
+        lambda: "huber_loss does not support non-positive values for delta.",
+    )
+    z = (input - target).abs()
+    loss = torch.where(z < delta, 0.5 * z * z, delta * (z - 0.5 * delta))
```
Comment (Lezcano) [inline torch/_refs/nn/functional/__init__.py:445]: Obs. This is fine for real numbers, as taking `abs` means we just multipyl by the sign. For complex numbers, it'd be better to do `diff = input - target` and then `diff * dif.conj()`. Now, this loss does not support complex numbers, so that's completely fine .

**Review by lezcano — APPROVED** [review]
Cool!

File: torch/_refs/nn/functional/__init__.py [hunk 2]
```diff
@@ -419,6 +420,32 @@ def hinge_embedding_loss(
...
+# the Python frontend and ATen:
+# https://github.com/pytorch/pytorch/issues/83931
+@elementwise_type_promotion_wrapper(
+    type_promoting_args=("input", "target"),
+    type_promotion_kind=ELEMENTWISE_TYPE_PROMOTION_KIND.DEFAULT,
+)
+def huber_loss(
+    input: TensorLikeType,
+    target: TensorLikeType,
+    reduction: str = "mean",
```
Comment (ngimel) [inline torch/_refs/nn/functional/__init__.py:433]: the only difference is in reduction argument? I think we can make reference accept both string and integer reduction argument.

File: torch/_refs/nn/functional/__init__.py [hunk 3]
```diff
@@ -419,6 +420,32 @@ def hinge_embedding_loss(
...
 
 
+# TODO: Cannot register as decomp here due to incompatible signatures between
+# the Python frontend and ATen:
+# https://github.com/pytorch/pytorch/issues/83931
+@elementwise_type_promotion_wrapper(
+    type_promoting_args=("input", "target"),
+    type_promotion_kind=ELEMENTWISE_TYPE_PROMOTION_KIND.DEFAULT,
+)
+def huber_loss(
```
Comment (ngimel) [inline torch/_refs/nn/functional/__init__.py:430]: do you need `out` wrapper here?

File: torch/_refs/nn/functional/__init__.py (hunk 1 above)
Comment (ngimel) [reply torch/_refs/nn/functional/__init__.py:445]: `complex_tensor.abs()` would do the right thing, no?

Comment (Lezcano) [reply torch/_refs/nn/functional/__init__.py:445]: For complex tensors, we would have an expression like `diff.abs()**2`, but you can save the `sqrt`  within `abs()` by writing it as `diff.abs() ** 2 == diff * diff.conj()`. But again, this loss does not support complex numbers in core, so that's fine.

File: torch/_refs/nn/functional/__init__.py (hunk 3 above)
Comment (mruberry) [reply torch/_refs/nn/functional/__init__.py:430]: Interestingly `nn.functional.huber_loss` doesn't support an `out` parameter, which is a UX consistency bug:

https://github.com/pytorch/pytorch/blob/d52452b3d1ff14a87e1d79c8d8bd67f0f074e6e6/torch/nn/functional.py#L3199

We should probably scrub and correct these UX consistency issues together. I made a note to follow-up.

File: torch/_refs/nn/functional/__init__.py (hunk 1 above)
Comment (mruberry) [reply torch/_refs/nn/functional/__init__.py:445]: We could add a comment about future possible complex number support

File: torch/testing/_internal/common_methods_invocations.py [hunk 4]
```diff
@@ -7009,6 +7009,18 @@ def sample_inputs_huber_loss(op_info, device, dtype, requires_grad, **kwargs):
...
         yield SampleInput(input, args=(target, ), kwargs=d)
 
+def error_inputs_huber_loss(op, device, **kwargs):
+    make_input = partial(make_tensor, device=device, dtype=torch.float32)
+    # invalid reduction value
+    err = 'is not a valid value for reduction'
+    yield ErrorInput(SampleInput(make_input(5, 4), args=(make_input(5, 4),), kwargs={'reduction': 'abc'}),
+                     error_type=ValueError, error_regex=err)
+    # delta <= 0
+    for delta in (0, -1):
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:7019]: Nice error checks

File: torch/_refs/nn/functional/__init__.py (hunk 2 above)
Comment (ngimel) [reply torch/_refs/nn/functional/__init__.py:433]: How about doing it the other way around, and have torch reference map a str argument to an int, and the call _decomp aten decomposition?

Comment (mruberry) [reply torch/_refs/nn/functional/__init__.py:433]: > How about doing it the other way around, and have torch reference map a str argument to an int, and the call _decomp aten decomposition?

It doesn't really matter which way it would be done today

File: torch/_refs/nn/functional/__init__.py (hunk 3 above)
Comment (nkaretnikov) [reply torch/_refs/nn/functional/__init__.py:430]: no, i think. the yaml file has the out overload, but the opinfo and the function signature say that `out` is not supported

Comment (Lezcano) [reply torch/_refs/nn/functional/__init__.py:430]: Note that this is an issue with many of the loss functions really: the function has an out variant on ATen, but it's not exposed on the `functional` API.

Comment (nkaretnikov) [reply torch/_refs/nn/functional/__init__.py:430]: > We should probably scrub and correct these UX consistency issues together. I made a note to follow-up.

note: i have a wip local branch that allows testing for `out` support consistency between regular pytorch ops and primtorch refs. it needs some clean up, but i'm planning to submit once the ref alias checking stack is merged

Comment (Lezcano) [reply torch/_refs/nn/functional/__init__.py:430]: Note that it's fine if a reference has the `out=` behaviour but the original operation doesn't. Also, we already have a number of skips/errors for `test_out` in core, so I'd expect that, if anything, the references are more correct than the functions in core. It may be a bit tricky to write a test that's not too flaky and that test all these issues.

**Comment by nkaretnikov** [comment]
@mruberry @ngimel addressed feedback, ci is green, ptal

**Review by mruberry — APPROVED** [review]
Cool!

**Comment by nkaretnikov** [comment]
@ngimel a gentle reminder that this one is waiting for review, PTAL

**Comment by ngimel** [comment]
This is approved, you can land it.
